cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -454,6 +456,172 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
   return Status::OK();
 }
 
+class HashJoinNode;
+
+struct BloomFilterPushdownContext {
+  using BuildFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  using FiltersReceivedCallback = std::function<Status()>;
+  using FilterFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  void Init(HashJoinNode* owner, size_t num_threads, TaskScheduler* scheduler,
+            FiltersReceivedCallback on_bloom_filters_received, bool disable_bloom_filter,
+            bool use_sync_execution);
+
+  Status StartProducing();
+
+  void ExpectBloomFilter() { eval_.num_expected_bloom_filters_ += 1; }
+
+  Status BuildBloomFilter(size_t thread_index, AccumulationQueue batches,
+                          BuildFinishedCallback on_finished);
+
+  Status PushBloomFilter();
```

Comment:
Given that this is a .cc file, why are some of these defined inline and some are only declared here?

Reply:
The things that need to access `HashJoinNode` need to be defined below `HashJoinNode`'s definition.